### PR TITLE
Transformer encoder

### PR DESCRIPTION
**Pull request by lutovatanya**

### CODE REVIEW

**Comment by lutovatanya** [comment]
image

Получилось большое число параметров, если я все сделала так

File: main.py [hunk 1]
```diff
@@ -1,12 +1,31 @@
...
 from torch.nn.modules.activation import MultiheadAttention
 from torch.nn.modules.dropout import Dropout
 from torch.nn.modules.linear import Linear
 from torch.nn.modules.normalization import LayerNorm
 from typing import Optional
+import math
+from torch.nn.init import xavier_uniform_, constant_, xavier_normal_
+from torch.nn.parameter import Parameter
+
+class PositionalEncoding(nn.Module):
```
Comment (alexeyhorkin) [inline main.py:16]: We don't use this class yet, but we have to include it into the final `TransformerModel` model

File: main.py [hunk 2]
```diff
@@ -41,9 +60,110 @@ def forward(self, src: Tensor, src_mask: Optional[Tensor] = None,
...
+                v_proj_weight=self.v_proj_weight)
+        else:
+            return F.multi_head_attention_forward(
+                query, key, value, self.embed_dim, self.num_heads,
+                self.in_proj_weight, self.in_proj_bias,
+                self.bias_k, self.bias_v, self.add_zero_attn,
+                self.dropout, self.out_proj.weight, self.out_proj.bias,
+                training=self.training,
+                key_padding_mask=key_padding_mask, need_weights=need_weights,
+                attn_mask=attn_mask)
```
Comment (alexeyhorkin) [inline main.py:147]: Let's remove it sicne we don't need this class at all. And we shouldn't rewrite `MultiheadAttention` class, we can simply import it from `torch` package as well.

File: main.py [hunk 3]
```diff
@@ -1,12 +1,33 @@
...
+        pe = torch.zeros(max_len, 1, d_model)
+        pe[:, 0, 0::2] = torch.sin(position * div_term)
+        pe[:, 0, 1::2] = torch.cos(position * div_term)
+        self.register_buffer('pe', pe)
+
+    def forward(self, x: Tensor) -> Tensor:
+        x = x + self.pe[:x.size(0)]
+        return self.dropout(x)
 
 class TransformerEncoderLayer(nn.Module):
```
Comment (alexeyhorkin) [inline main.py:32]: Let's also delete `TransformerEncoderLayer` since we simply can import it from torch package.

File: main.py [hunk 4]
```diff
@@ -41,9 +62,127 @@ def forward(self, src: Tensor, src_mask: Optional[Tensor] = None,
...
+    print(out2)
+
+
+    multihead_attn = nn.MultiheadAttention(1200, 8)
+    q = torch.randn(10, 1, 1200)
+    k = torch.randn(10, 1, 1200)
+    v = torch.randn(10, 1, 1200)
+    attn_output, attn_output_weights = multihead_attn(q,k,v)
+    print(attn_output.shape, attn_output_weights.shape)
+
```
Comment (alexeyhorkin) [inline main.py:187]: Let's remove it, since we don't have to test `nn.MultiheadAttention`!

File: main.py [hunk 5]
```diff
@@ -41,9 +62,127 @@ def forward(self, src: Tensor, src_mask: Optional[Tensor] = None,
...
+
+if __name__ == "__main__":
+
+    encoder_par = count_parameters(nn.TransformerEncoderLayer(d_model=1200, nhead=8))
+    print(encoder_par)
+
+    encoder_layer = nn.TransformerEncoderLayer(d_model=1200, nhead=8)
+    print(encoder_par)
+    src1 = torch.randn(10, 1, 1200)
```
Comment (alexeyhorkin) [inline main.py:169]: Please, pay attention, that `nn.TransformerEncoderLayer` expects input in the following format shape: `[H, BS, C]`, for us `H` = 1200, `BS` = 10, `C`=1

File: main.py [hunk 6]
```diff
@@ -41,9 +62,127 @@ def forward(self, src: Tensor, src_mask: Optional[Tensor] = None,
...
+    print(src1)
+    out1 = encoder_layer(src1)
+    print(out1)
+
+    encoder_layer = nn.TransformerEncoderLayer(d_model=1400, nhead=8)
```
Comment (alexeyhorkin) [inline main.py:174]: Please, pay attention, that `nn.TransformerEncoderLayer` expects input in the following format shape: `[H, BS, C]`, for us `H` = 1400, `BS` = 10, `C`=1

**Review by alexeyhorkin — COMMENTED** [review]
In general, we have to add smth like `TransformerModel` class, which would contains Pos encoding + `nn.TransformerEncoder` + final `nn.Linear(in_dim, 4)`. Linear layer needs for the final logits. And also, pay attention that all these torch layers operate with `[H, BS, C]` shape. So, we can do forward pass in `TransformerModel` model the same way for now, but then, we will use [permute](https://pytorch.org/docs/stable/generated/torch.permute.html) function while forward passing.

File: main.py (hunk 5 above)
Comment (lutovatanya) [reply main.py:169]: image

Я ориентировалась на этот пример в документации

Если изменить порядок, будет ошибка 
image
 
 Получается, тогда выше нужно
 encoder_layer = nn.TransformerEncoderLayer(d_model=1, nhead=1) 
   вот так?

Comment (alexeyhorkin) [reply main.py:169]: Ага:
```python3
d_model = 1
nhead = 1
dropout = 0.5
d_hid = 128  # например так для начала
nn.TransformerEncoderLayer(d_model, nhead, d_hid, dropout)
```
